HEVC 10bit encoder and P010/I010 in put enable

QuPengfei opened the pull request:
this enable the P010/I010 format support and HEVC 10bit encoder

Commit:
HEVC ENC:10bit support

add the P010/I010 input
add support for HEVC 10bit encoding

Commit:
HEVC ENC:10bit support

add the P010/I010 input
add support for HEVC 10bit encoding

Commit:
Merge branch 'apache' of https://github.com/QuPengfei/libyami into hevc10bit

xuguangxin commented on vaapi/vaapisurfaceallocator.cpp:
```diff
@@ -33,24 +33,35 @@ VaapiSurfaceAllocator::VaapiSurfaceAllocator(VADisplay display, uint32_t extraSi
 YamiStatus VaapiSurfaceAllocator::doAlloc(SurfaceAllocParams* params)
 {
+    uint32_t rtFormat = VA_RT_FORMAT_YUV420;
+    VASurfaceAttrib attrib;
```
Better memset any va structure before we use it, It maybe add fields in future.

xuguangxin replied:
Could we move this to line 58?

QuPengfei replied:
agree to memset.  it is ok to move to line 58.

xuguangxin commented on interface/VideoCommonDefs.h:
```diff
@@ -69,6 +69,8 @@ extern "C" {
 
 #define YAMI_FOURCC_NV12 YAMI_FOURCC('N','V','1','2')
+#define YAMI_FOURCC_P010 YAMI_FOURCC('P','0','1','0')
+#define YAMI_FOURCC_I010 YAMI_FOURCC('I','0','1','0')
```
Do we really need this in core library level. Could we support P010 only? And let yami-utils do I010 to P010 conversion?

QuPengfei replied:
it is OK to only use P010. For I010, it is OK to use by YAMI_FOURCC('I','0','1','0')

xuguangxin commented on encoder/vaapiencoder_hevc.cpp:
```diff
@@ -866,6 +867,7 @@ VaapiEncoderHEVC::VaapiEncoderHEVC():
 {
     m_videoParamCommon.profile = VAProfileHEVCMain;
     m_videoParamCommon.level = 51;
+    m_videoParamCommon.bitDepth = 8;
```
better put this to base class. Every codec need this. Actually 

> m_pool = SurfacePool::create(m_alloc, ((m_videoParamCommon.bitDepth == 8)?YAMI_FOURCC_NV12:YAMI_FOURCC_P010), (uint32_t)surfaceWidth, (uint32_t)surfaceHeight, m_maxOutputBuffer);

will wrong if we do not set this in base

lizhong1008 replied:
Agree with @xuguangxin

QuPengfei replied:
agree

xuguangxin commented on encoder/vaapiencoder_base.cpp:
```diff
@@ -498,7 +661,7 @@ bool VaapiEncoderBase::initVA()
     int32_t surfaceWidth = ALIGN16(m_videoParamCommon.resolution.width);
     int32_t surfaceHeight = ALIGN16(m_videoParamCommon.resolution.height);
-    m_pool = SurfacePool::create(m_alloc, YAMI_FOURCC_NV12, (uint32_t)surfaceWidth, (uint32_t)surfaceHeight, m_maxOutputBuffer);
+    m_pool = SurfacePool::create(m_alloc, ((m_videoParamCommon.bitDepth == 8)?YAMI_FOURCC_NV12:YAMI_FOURCC_P010), (uint32_t)surfaceWidth, (uint32_t)surfaceHeight, m_maxOutputBuffer);
```
Could we return false for other bitDepth?

lizhong1008 replied:
why forcc is P010 if is not 8 bits? it can be VA_FOURCC_P016.

QuPengfei replied:
feed the VA_FOURCC_P016 when 10bit
use VA_FOURCC_NV12 as the default mode when bitDepth !=8

and if we add 12bit or more ,just add here in the later

xuguangxin commented on common/utils.cpp:
```diff
@@ -130,13 +131,24 @@ bool getPlaneResolution(uint32_t fourcc, uint32_t pixelWidth, uint32_t pixelHeig
+                height[1] = (h + 1) >> 1;
+                planes = 2;
+            } else if (fourcc == VA_FOURCC('I', '0', '1', '0')) {
+                width[1] = width[2] = (w + 1) >> 1;
```
little confused, what is the diffrence between I420 and I010, why we can't merge this with else branch.Seems they are same code

lizhong1008 replied:
For 10bits pixel, it should need 16bits with the lowest 6 bits set to zero. so,width[0]= 2*w,  width[1] = width[2] = w, right?

xuguangxin replied:
I mean seems L149 to L152 is duplicate for L153 to L155, why we can't merge them to else branch?

lizhong1008 replied:
L149 to L 151 is error, I010 is differenct from I420.

QuPengfei replied:
zhong is correct. wrong width calculation for I010 here. 

```
        } else if (fourcc == VA_FOURCC('I', '0', '1', '0')) {
            width[0] = 2* w; 
            width[1] = width[2] = w;
            height[1] = height[2] = (h + 1) >> 1;
            planes = 3;
        } else {
```

QuPengfei replied:
there is error when NV12
            if (fourcc == VA_FOURCC_NV12) {
                 width[1]  = w + (w & 1);
                 height[1] = (h + 1) >> 1;
                 planes = 2;
here width[1] = w;

xuguangxin commented on encoder/vaapiencoder_base.cpp:
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
+    return bpp;
+}
+
+static bool loadFrameToSurface(VAImage surface_image,uint8_t* dst,uint8_t* src,uint32_t fourcc)
```
why we need this? could handle it with getPlaneResolution and copyImage?

QuPengfei replied:
it convert I010->P010 and I420->NV12. copyImage doesnot.

the driver can convert I420->NV12 when input is I420.  and it does not support I010->I420.

xuguangxin commented:
Hi @QuPengfei , thanks for your patch. We should suggest other member to implement new features like this; it will save efforts for all of us. ;).

lizhong1008 commented:
Seems need to change these commit names.

lizhong1008 commented on encoder/vaapiencoder_base.cpp:
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
+    case VA_FOURCC_I420:
+        bpp = 1;
+        break;
+
```
line 356 to 360 can be removed.

QuPengfei replied:
it is better keep the codes there. it help.

lizhong1008 commented on encoder/vaapiencoder_base.cpp:
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
+            return false;
+        }
+
+        y_dst = (unsigned char *)((unsigned char*)surface_p + surface_image.offsets[0]);
```
would better use uint8\* instead of unsigned char *. And the first (unsigned char *) is not necessary.

QuPengfei replied:
agree

lizhong1008 commented on encoder/vaapiencoder_base.cpp:
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
+            v_src = src + surface_image.width * surface_image.height * 5 / 4 * bpp;
+        } else {
+            ERROR("only support I010/I420 ");
+            return false;
```
We can return false if fourcc is not I010 or I420 at the first of this function to make code more clear.

QuPengfei replied:
agree

lizhong1008 commented on encoder/vaapiencoder_base.cpp:
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
+            ERROR("don't support this surface fcc");
+            return false;
+        }
+
```
No need these code.

lizhong1008 commented on encoder/vaapiencoder_base.cpp:
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
+
+        /* Y plane, directly copy */
+        for (row = 0; row < surface_image.height; row++) {
+            if(bpp == 2){
```
would better use forcc ==10bits,  instead of bpp==2. If one day we need to support 16 bits pixel(here VA_FOURCC_P016 is added), the bpp still equal to 2, but the logic is different

xuguangxin replied:
Not sure why we need loadFrameToSurface. Could we just use getPlaneResolution and copyImage like other formats?

QuPengfei replied:
@lizhong1008, from my understanding, the 16 bit  has the same logic as 10bits. i am little confused what is the difference here?

lizhong1008 replied:
@QuPengfei Dose 16 bits need "value <<= 6" in line 426?

QuPengfei replied:
yes, 16bits does not need this line. here is a bug.

lizhong1008 commented:
Does this patch series support hevc 10 bits encoding? seems bitDepth is always 8, and profile is always HEVC main profile, not Main 10 profile.

chivakker commented on encoder/vaapiencoder_base.cpp:
```diff
@@ -274,16 +274,23 @@ SurfacePtr VaapiEncoderBase::createNewSurface(uint32_t fourcc)
     uint32_t rtFormat;
     SurfacePtr surface;
 
+
```
unnecessary change

QuPengfei replied:
agree

chivakker commented on vaapi/vaapisurfaceallocator.cpp:
```diff
@@ -33,24 +33,35 @@ VaapiSurfaceAllocator::VaapiSurfaceAllocator(VADisplay display, uint32_t extraSi
-    if (params->fourcc != YAMI_FOURCC_NV12) {
-        ERROR("fix me for 10 bits");
+    if (params->fourcc != YAMI_FOURCC_NV12 && params->fourcc != YAMI_FOURCC_P010) {
+        ERROR("only support NV12 and P010");
```
Needs clang format

QuPengfei replied:
agree

QuPengfei commented:
thanks for your comments. guanxing,lizhong and Daniel .

Commit:
FIX:I010 byte width caculation incorrectly.

fix the following
1.I010 caculate wrong byte width
2.memset for attribte member
3.other changes according to review comments

lizhong1008 commented:
@QuPengfei ,  any update for it?

QuPengfei commented:
According to code reivew, the first version changes are reday on https://github.com/QuPengfei/libyami.

i also talk with guangxin last week. he hope to upstream JPEG related changes firstly. and then the patch still need modification based on it.

what is your plan for it?

lizhong1008 commented:
@QuPengfei , "the  JPEG related changes (https://github.com/01org/libyami/pull/537)" have been merged. You can rebased on it.

lizhong1008 commented:
@QuPengfei , any update for this pull request?

xuguangxin commented:
this will rebase after my hevc decoder branch merged. Since many things have been changed.

xuguangxin commented:
Hi @QuPengfei 
After some code refact, add p010 is pretty simply compare to previous patch. It only need one line code https://github.com/01org/libyami/pull/638/commits/788be3ba30e68c9332314be5165f239720e63f63. 
I have add to my hevc 10 bit decoder,Please rebase this on my patch set.
thanks

xuguangxin commented:
close this since we have #668
